### ISSUE TITLE
Prevent Sending Passwords from the Clipboard

### ISSUE THREAD

**Issue by cesar19004** [GSConnect/gnome-shell-extension-gsconnect#1893]
### Describe your request

Auto-sharing the clipboard to a mobile device is really useful; however, when using password managers, you might not want to share passwords, as they could be stored insecurely in the mobile device's clipboard manager.

### Proposed solution

Use clipboard MIME data to determine if content should be shared. Most password managers set the appropriate MIME data to mark content as private, as demonstrated in this [example](https://github.com/keepassxreboot/keepassxc/blob/bf856d278d21be2e54d5590fb7e69e63184262d3/src/gui/Clipboard.cpp#L57-L63) from KeePassXC.

KDE Connect already [uses](https://invent.kde.org/network/kdeconnect-kde/-/blob/master/plugins/clipboard/clipboardlistener.cpp#L64) the `x-kde-passwordManagerHint` MIME data to [check for passwords](https://invent.kde.org/network/kdeconnect-kde/-/blob/master/plugins/clipboard/clipboardplugin.cpp#L40-44). It also includes a [config option](https://invent.kde.org/network/kdeconnect-kde/-/blob/master/plugins/clipboard/clipboardplugin.cpp#L49-L60) to control whether sharing passwords is permitted.

### GSConnect version

58

### Installed from

GNOME Extensions website

### GNOME Shell version

47

### Linux distribution/release

Ubuntu 24.10

**Comment by cesar19004** [comment]
Thanks for taking the time.
If I'm correct, these are the MIME types accepted by GSConnect:
```js
const TEXT_MIMETYPES = [
    'text/plain;charset=utf-8',
    'UTF8_STRING',
    'text/plain',
    'STRING',
];
```

And these are the MIME types of passwords returned by the GNOME evaluator when running `St.Clipboard.get_default().get_mimetypes(Meta.SelectionType.SELECTION_CLIPBOARD)`:
```
0: text/plain;charset=utf-8
1: x-kde-passwordManagerHint
2: text/plain
```

For now, maybe a simple validation to check that `x-kde-passwordManagerHint` is not present in the MIME types should be enough, while an option in the config to control password sharing is being considered.

**Comment by ferdnyc** [comment]
@cesar19004 

Mmm, it looks like currently we special-case ignore any clipboard with `text/uri-list` in its mimetypes... 

https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/734ef1ec3cf531d78b481f9cf32f7e09f4e59258/src/service/components/clipboard.js#L182-L193

I don't see why we couldn't extend that to also exclude `x-kde-passwordManagerHint`... although it maybe seems a bit odd to see `x-kde-passwordManagerHint` clipboard contents in a GNOME Shell session? (Still no harm in doing it, tho.)

**Comment by ferdnyc** [comment]
My one slight worry with this is whether there might be users who _rely_ on copy-pasting passwords to their mobile device, who we might be breaking things for if we disable this across-the-board.

**Comment by cesar19004** [comment]
Agreed, using `x-kde-passwordManagerHint` in GNOME Shell seems odd. Also, it doesn't seem to be widely adopted at the moment and passwords copied from some password managers could still be shared. However, in the meantime, it shouldn’t cause harm to validate it while we wait for a more standard or better alternative.

**Comment by cesar19004** [comment]
Perhaps, to avoid breaking the copy-pasting of passwords for users who use it, it could be added as an experimental option disabled by default

**Comment by cesar19004** [comment]
Hi @daniellandau,
Yes, by setting that option to be disabled by default, I meant that it would only prevent sending passwords when the option is enabled.

> I think it would be weird for us to set an x-kde mimetype, but aren't we just talking about handling that when set by others? KeepassXC is definitely used in gnome desktops.

Yeah, I see no reason to set that MIME type, only to read it. KeepassXC is quite popular, so I agree that it's definitely used by GNOME users.

**Comment by ferdnyc** [comment]
@daniellandau 

> I would think having a trusted password manager in one device and copy pasting passwords over gsconnect to another is a valid, perhaps even quite normal use case. So I wouldn't make it the default to exclude those. Saves on having to figure out syncing a keepass database for example and dealing with sync conflicts.

Mmm, that's fair. So I guess we do need to place password-blocking behind an option, disabled by default, but there for those who want extra protection.

> I think it would be weird for us to set an x-kde mimetype, but aren't we just talking about handling that when set by others? KeepassXC is definitely used in gnome desktops.

Yeah, I suppose they set a KDE identifier simply because KDE defined one for that purpose, and GNOME (presumably) never did.

@cesar19004 
> Yeah, I see no reason to set that MIME type, only to read it.

Well, I _am_ kind of curious what happens on the Android side, if you try to copy a password _to_ the paired device. Will Android even allow KDE Connect to share passwords in that direction? Or will it block the sharing of sensitive data, the same way Android 15 has cut off our access to OTP strings in SMS notifications? (See https://github.com/GSConnect/gnome-shell-extension-gsconnect/discussions/1833#discussioncomment-11091768)

**IF** it _is_ possible for KDE Connect to send us passwords over the clipboard, and **IF** we have a way to identify them as such, it'd probably be a good idea for us to add the `-kde-passwordManagerHint` mimetype when we publish them to the GNOME clipboard.
